File: src/onjava/codility/primencomposite/Peaks.java
package onjava.codility.primencomposite;

import org.testng.Assert;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;


/**
 * Site: Codility
 * Problem: Peaks
 *
 * User: Luis Villaca
 * Date: 24/05/2020
 */
public class Peaks {
    public static int solution(int[] A) {
        //Peaks
        int lenghtOfA = A.length;
        int[] peaks = new int[lenghtOfA];

        if (lenghtOfA < 3) return 0;

        peaks[0] = 0;

        for (int n = 1; n < lenghtOfA - 1; n++) {
            if (A[n - 1] < A[n] && A[n] > A[n + 1]) {
                peaks[n] = peaks[n-1]+1;   // accumulating peaks
            } else {
                peaks[n] = peaks[n-1];
            }
        }


        // updating last line (not included above)
        peaks[lenghtOfA - 1] = peaks[lenghtOfA - 2];

        int nbrPeaks = peaks[lenghtOfA-1];

        if (nbrPeaks < 1) return 0;
        if (nbrPeaks == 1) return 1;

        int maxDivs = 1;

        for (int n = 1; n <= nbrPeaks; n++) {
            if (lenghtOfA % n == 0) { // divide equally

                //first block
                int blockSize = lenghtOfA / n;
                int prevBlock = 0;
                int endBlock = blockSize - 1;
                boolean terminate=false;

                while (endBlock< lenghtOfA) {
                    if (!hasPeak(prevBlock, endBlock, peaks)) {
                        terminate=true;
                        break;
                    }
                    prevBlock = endBlock + 1;
                    endBlock += blockSize;
                }

                if (terminate) {
                    continue;
                } else {
                    maxDivs= n;
                }
            }
        }

        return maxDivs;
    }

    public static boolean hasPeak(int init, int end, int[] peaks) {
        //Peaks
        if (init == 0) return peaks[end] > 0;
        else {
            return peaks[end] > peaks[init - 1];
        }
    }


    @DataProvider(name = "test")
    public Object[][] createData1() {
        return new Object[][]{
                new Object[]{new int[]{1, 2, 3, 4, 3, 4, 1, 2, 3, 4, 6, 2}, 3},
                new Object[]{new int[]{1, 3, 2, 1}, 1},
                new Object[]{new int[]{1, 4, 1, 3, 2}, 1},
                new Object[]{new int[]{1, 2, 3, 4, 3, 4, 1, 2, 3, 4, 6, 2, 1, 2, 5, 2}, 4},
                new Object[]{new int[]{0, 1, 0, 1, 0, 1, 0, 1}, 2},
                new Object[]{new int[]{0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1}, 4},
                new Object[]{new int[]{0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1}, 4},
                new Object[]{new int[]{0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0}, 4},
                new Object[]{new int[]{0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0}, 4},
                new Object[]{new int[]{0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0}, 4},
                new Object[]{new int[]{0, 1, 0, 0, 1, 0, 0, 1, 0}, 3},
                new Object[]{new int[]{0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0}, 1},
                new Object[]{new int[]{1, 2, 3, 4, 5, 6, 5, 5, 5, 6, 5, 6}, 2},
                new Object[]{new int[]{1, 2, 1, 1, 2, 1, 1, 2, 1, 1, 2, 1, 1, 2, 1, 1, 2, 1, 1, 2, 1, 1, 2, 1, 1, 2, 1, 1, 2, 1}, 10},
                new Object[]{new int[]{5}, 0},
        };
    }

    @Test(dataProvider = "test")
    public void verifySolution(int[] pInput, int expectedCount) {
        Assert.assertEquals(solution(pInput), expectedCount);
    }
}

File: src/onscala/codility/primencomposite/Peaks.scala
package onscala.codility.primencomposite

import org.scalatest.Assertions
import org.testng.Assert._
import org.testng.annotations.{DataProvider, Test}

import scala.annotation.tailrec

/**
  * Site: Codility
  * Problem: Peaks 
  *
  * User: Luis Villaca
  * Date: 24/05/2020
  */
class Peaks extends Assertions {

  def solution(a: Array[Int]): Int  = {

    val arrLength = a.length
    if (arrLength < 3) 0
    else {
      // 1 - aux accumulated array -----------------------------------------
      val accArray = new Array[Int](arrLength)

      // initial index
      accArray(0) = 0

      (1 to arrLength-2).foreach(internalIndex => {
        if (a(internalIndex-1) < a(internalIndex) && a(internalIndex) > a(internalIndex+1)){
          // a peek
          accArray(internalIndex) = accArray(internalIndex-1) +1;
        } else {
          accArray(internalIndex) = accArray(internalIndex-1);
        }
      })

      // last index
      accArray(arrLength-1) = accArray(arrLength-2)
      val nbrPeaks = accArray(arrLength-1)

      // 2 - helper functions  -----------------------------------------


      val hasPeakInBlock = (init: Int, end: Int) => {
        if (init == 0) accArray(end) > 0
        else accArray(end) > accArray(init - 1)
      }

      val hasPeakInBlocks = (factor: Int) => {
        val blockSize = arrLength / factor;
        def hasPeakInBlocksRec(initBlock: Int, endBlock: Int, maxAcc:Int): Int = {
          if (endBlock > arrLength) maxAcc
          else if (!hasPeakInBlock(initBlock, endBlock)) 0
          else hasPeakInBlocksRec(initBlock+blockSize, endBlock+blockSize, maxAcc+1)
        }
        val initBlock = 0
        val endBlock =  blockSize - 1
        hasPeakInBlocksRec(initBlock, endBlock, 0) == factor
      }

      val isFactor = (x:Int) => arrLength % x ==0

      // 3 - get max factor that has all peaks ---------------------------

      (1 to nbrPeaks).filter(isFactor) //is factor
        .filter(hasPeakInBlocks) // map to block size
        .reduceOption(_ max _).getOrElse(0)
    }

  }


  //Evaluating
  @DataProvider(name= "tests")
  def testData() = Array(
    Array[Any](Array(1, 2, 3, 4, 3, 4, 1, 2, 3, 4, 6, 2), 3),
    Array[Any](Array(1, 3, 2, 1), 1),
    Array[Any](Array(1, 4, 1, 3, 2), 1),
    Array[Any](Array(1, 2, 3, 4, 3, 4, 1, 2, 3, 4, 6, 2, 1, 2, 5, 2), 4),
    Array[Any](Array(0, 1, 0, 1, 0, 1, 0, 1), 2),
    Array[Any](Array(0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1), 4),
    Array[Any](Array(0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1), 4),
    Array[Any](Array(0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0), 4),
    Array[Any](Array(0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0), 4),
    Array[Any](Array(0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0), 4),
    Array[Any](Array(0, 1, 0, 0, 1, 0, 0, 1, 0), 3),
    Array[Any](Array(0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0), 1),
    Array[Any](Array(1, 2, 3, 4, 5, 6, 5, 5, 5, 6, 5, 6), 2),
    Array[Any](Array(1, 2, 1, 1, 2, 1, 1, 2, 1, 1, 2, 1, 1, 2, 1, 1, 2, 1, 1, 2, 1, 1, 2, 1, 1, 2, 1, 1, 2, 1), 10),
    Array[Any](Array(5), 0),
  )

  @Test(dataProvider = "tests")
  def verify(arrayInput: Array[Int], expectedOutput: Int): Unit = {
    val actualOutput = solution(arrayInput)
    assertEquals(actualOutput, expectedOutput)
  }

}







